pi.py

```diff
@@ -1 +1,578 @@
+            TGNEW=T[j]
+            TJC=T[j]-273.15
+            ES=6.112*np.exp(17.67*TJC/(243.5+TJC))
+            RG=EPS*ES/(P[j]-ES)
+            
+            #
+            #   ***  Iteratively calculate lifted parcel temperature and mixing   ***
+            #   ***                ratio for reversible ascent                    ***
+            #
+            
+            # set loop counter and initial condition
+            NC=0
+            TG=0
+
+            # loop until loop converges or bails out
+            while ((np.abs(TGNEW-TG)) > 0.001):
+            
+                # Parcel temperature and mixing ratio during this iteration
+                TG=TGNEW
+                TC=TG-273.15
+                ENEW=6.112*np.exp(17.67*TC/(243.5+TC))
+                RG=EPS*ENEW/(P[j]-ENEW)
+                
+                # increase iteration count in the loop
+                NC=NC+1
+                
+                #
+                #   ***  Calculate estimates of the rates of change of the entropy    ***
+                #   ***           with temperature at constant pressure               ***
+                #
+
+                ALV=ALV0+CPVMCL*TC
+                SL=(CPD+RP*CL+ALV*ALV*RG/(RV*TG*TG))/TG
+                EM=RG*P[j]/(EPS+RG)
+                # (last term vanishes with saturation, i.e. RH=1)
+                SG=(CPD+RP*CL)*np.log(TG)-RD*np.log(P[j]-EM)+ALV*RG/TG
+                # convergence speed (AP, step in entropy fraction) varies as a function of 
+                # number of iterations
+                if (NC < 3):
+                    # converge slowly with a smaller step
+                    AP=0.3
+                else:
+                    # speed the process with a larger step when nearing convergence
+                    AP=1.0
+                # find the new temperature in the iteration
+                TGNEW=TG+AP*(S-SG)/SL
+                
+                #
+                #   ***   If the routine does not converge, set IFLAG=2 and bail out   ***
+                #
+                if (NC > 500) or (ENEW > (P[j]-1)):
+                    CAPED=0
+                    TOB=T[0]
+                    LNB=P[0]
+                    IFLAG=2
+                    # Return the uncoverged values
+                    return(CAPED,TOB,LNB,IFLAG)
+                
+                # store the number of iterations
+                NCMAX=NC
+                
+            #
+            #   *** Calculate buoyancy   ***
+            #
+            # Parcel Mixing ratio: either reversible (ascent_flag=0) or pseudo-adiabatic (ascent_flag=1)
+            RMEAN=ascent_flag*RG+(1-ascent_flag)*RP
+            # Parcel and Environmental Density Temperatures at this pressure (E94, EQN. 4.3.1 and 6.3.7)
+            TLVR=TG*(1.+RG/EPS)/(1.+RMEAN)
+            TENV=T[j]*(1.+R[j]/EPS)/(1.+R[j])
+            # Bouyancy of the parcel in the environment (Proxy of E94, EQN. 6.1.5)
+            TVRDIF[j,]=TLVR-TENV
+            
```

Comment:
I might be missing something but up until this point, I think this is pretty easily vectorized (meaning you could input a 3-D matrix, [lat, lon, pressure]), and have PI calculated everywhere. This is a feature that would make me use the code (for my ensemble forecasting). One of our priorities is to look at the internal variability of PI, which currently we do not take into account; we just keep PI constant at t=0.

This could accomplished by storing a boolean mask where all the points have converged already. It would reduce the readability of the code but I think speed up the convergence. So, essentially, you will be doing Newton's gradient descent method but calculating gradients and tolerances using a matrix.

Also, you might have a look at scipy.optimize for nice optimization functions.

Reply:
This is an interesting idea, I like the possibility of deriving Tvdiff ahead of time.

Because this is a bit outside the scope of what I need the function for, and I really want to maintain the readability, I wouldn't be able to work on this as the lead. If you are interested in working together to create a branch that uses this method following your vision, I would definitely be on board.